narwhal/primary/src/core.rs

```diff
@@ -185,327 +173,260 @@ impl Core {
+                Err(status) => {
+                    if status.status() == anemo::types::response::StatusCode::BadRequest {
+                        return Err(DagError::NetworkError(format!(
+                            "unrecoverable error requesting vote for {header}: {status:?}"
+                        )));
+                    }
+                    missing_parents = Vec::new();
                 }
             }
-        }
 
-        for header in latest_headers {
-            if let Err(err) = self.process_header(&header).await {
-                error!("error on recovery path processing latest header: {:?}", err);
-            }
-        }
+            // Retry delay. Using custom values here because pure exponential backoff is hard to
+            // configure without it being either too aggressive or too slow. We want the first
+            // retry to be instantaneous, next couple to be fast, and to slow quickly thereafter.
+            tokio::time::sleep(Duration::from_millis(match attempt {
+                1 => 0,
+                2 => 100,
+                3 => 500,
+                4 => 1_000,
+                5 => 2_000,
+                6 => 5_000,
+                _ => 10_000,
+            }))
+            .await;
+        };
 
-        self
+        // Verify the vote.
+        ensure!(
+            header.round == vote.round,
+            DagError::InvalidRound {
+                expected: header.round,
+                received: vote.round
+            }
+        );
+        ensure!(
+            vote.digest == header.digest()
+                && vote.origin == header.author
+                && vote.round == header.round,
+            DagError::UnexpectedVote(vote.digest)
+        );
+        vote.verify(&committee)?;
+        Ok(vote)
     }
 
     #[instrument(level = "debug", skip_all, fields(header_digest = ?header.digest()))]
-    async fn process_own_header(&mut self, header: Header) -> DagResult<()> {
-        if header.epoch < self.committee.epoch() {
+    async fn propose_header(
+        name: PublicKey,
+        committee: Committee,
+        header_store: Store<HeaderDigest, Header>,
+        certificate_store: CertificateStore,
+        signature_service: SignatureService<Signature, { crypto::DIGEST_LENGTH }>,
+        metrics: Arc<PrimaryMetrics>,
+        network: anemo::Network,
+        header: Header,
+        mut cancel: oneshot::Receiver<()>,
+    ) -> DagResult<Certificate> {
+        if header.epoch < committee.epoch() {
             debug!("Proposer outdated");
-            return Ok(());
+            return Err(DagError::InvalidEpoch {
+                expected: committee.epoch(),
+                received: header.epoch,
+            });
         }
 
-        // Update the committee now if the proposer already did so.
-        self.try_update_committee().await;
+        // Process the header.
+        // DO NOT MERGE: Any need to verify or protect against equivocation from self-produced
```

Comment:
The equivocation protection exists in the proposer https://github.com/MystenLabs/sui/blob/49a283022d03bd1a1178b8c21c2c343bf31f22cc/narwhal/primary/src/proposer.rs#L172

so unless we want to do something additional here, we do have guarantees from the proposer that it will not re-send a different header for proposal for the same round